README.md

```diff
@@ -98,6 +98,152 @@ allows us to do:
+the protocol have taken place. This is a complicated thing to do statically in
+most programming languages, but in Rust it is easy because of _move semantics_.
+
+Using move semantics, we ensure that "taking a step" in the protocol (sending,
+receiving, etc) cannot be repeated.
+
+## Protocol types
+
+Any session-typed communication protocol is constructed from some basic
+building blocks. This section goes through them pair-wise, showing an action
+and its dual.
+
+A session-typed channel is defined as `Chan<E, P>` where `E` is an
+_environment_ and `P` is a protocol. The environment `E` is always `()` for
+newly created channels (ie it is empty).
+
+### `Eps`
+
+This is the final step of any terminating protocol. The simplest example is:
+
+    type Close = Eps;
+
+Any channel whose type is `Chan<E, Eps>` implements the `close()` function that
+closes the connection.
+
+### `Send` and `Recv`
+
+These are the most basic of actions: Transmitting and receiving values. The
+opposite of a `Send` with some type `T` is a `Recv` of type `T`.
+
+    type S = Send<String, Eps>;
+    type R = Recv<String, Eps>; // <S as HasDual>::Dual
+
+A channel of type `Chan<E, Send<T, P>>` implements the function `send(T) → Chan<E, P>`.
+
+A channel of type `Chan<E, Recv<T, Q>>` implements the function `recv() → (T, Chan<E, Q>)`.
+
+### `Choose` and `Offer`
+
+There is the option of making choices in the protocol, for one process to
+inform the other of a decision. The `Choose` and `Offer` constructs model such
+choices.
+
+    type C = Choose<Send<String, Eps>, Recv<String, Eps>>;
+    type O = Offer<Recv<String, Eps>, Send<String, Eps>; // <C as Hasdual>::Dual
+
+A channel of type `Chan<E, Choose<P, Q>>` implements _two_ functions:
+
+ * `sel1() → Chan<E, P>`
+ * `sel2() → Chan<E, Q>`
+
+that communicates the choice of protocols `P` and `Q` to the other process.
+
+A channel of type `Chan<E, Offer<P, Q>>` implements `offer() → Branch<Chan<E,
+P>, Chan<E, Q>>`. `Branch` is an enum defined as:
+
+    enum Branch<L, R> {
+        Left(L),
+        Right(R),
+    }
+
+A call to `offer()` should then be matched on to figure out which path the
+other process decided to take.
+
+    match c.offer() {
+        Branch::Left(c) => …,
+        Branch::Right(c) => …,
+    }
+
+### `Rec`, `Var`, `S` and `Z`
+
+The type `Rec` implements the ability the recurse in the protocol, ie provides
+an iterator component. The type `Rec<P>` allows repeating the protocol `P`.
+
+A channel of type `Chan<E, Rec<P>>` implements the function `enter() → Chan<(P,
+E), P>`. What `enter()` does is "store" the protocol `P` in the channel
+environment.
+
+The `Var` construct is then used to reference protocols stored in the
+environment. As the environment is essentially a stack, `Var` takes a counter
+that is modeled as a Peano number. `Var<Z>` points to the top of the stack.
+
+A channel of type `Chan<(P, E), Var<Z>>` implements the function `zero() →
```

Comment:
Again I think we should be clear that `zero` is a method. And perhaps it would be even more beneficial for the user if it said `zero(self)` (here and in other examples) thereby making it clear that the channel is consumed.

Reply:
Done.